pkg/subctl/cmd/cloud/gcp/gcp.go

```diff
@@ -0,0 +1,181 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This package provides common functionality to run cloud prepare/cleanup on GCP Clusters
+package gcp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+	"github.com/submariner-io/admiral/pkg/util"
+	"github.com/submariner-io/cloud-prepare/pkg/api"
+	cloudpreparegcp "github.com/submariner-io/cloud-prepare/pkg/gcp"
+	gcpClientIface "github.com/submariner-io/cloud-prepare/pkg/gcp/client"
+	"github.com/submariner-io/cloud-prepare/pkg/k8s"
+	"github.com/submariner-io/cloud-prepare/pkg/ocp"
+	cloudutils "github.com/submariner-io/submariner-operator/pkg/subctl/cmd/cloud/utils"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/cmd/utils"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/cmd/utils/restconfig"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/dns/v1"
+	"google.golang.org/api/option"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	infraIDFlag   = "infra-id"
+	regionFlag    = "region"
+	projectIDFlag = "project-id"
+)
+
+var (
+	infraID         string
+	region          string
+	projectID       string
+	credentialsFile string
+	ocpMetadataFile string
+)
+
+// AddGCPFlags adds basic flags needed by GCP
+func AddGCPFlags(command *cobra.Command) {
+	command.Flags().StringVar(&infraID, infraIDFlag, "", "GCP infra ID")
+	command.Flags().StringVar(&region, regionFlag, "", "GCP region")
+	command.Flags().StringVar(&projectID, projectIDFlag, "", "GCP project ID")
+	command.Flags().StringVar(&ocpMetadataFile, "ocp-metadata", "",
+		"OCP metadata.json file (or directory containing it) to read GCP infra ID and region from (takes precedence over the flags)")
+
+	dirname, err := os.UserHomeDir()
+	if err != nil {
+		utils.ExitOnError("failed to find home directory", err)
+	}
+
+	defaultCredentials := filepath.FromSlash(fmt.Sprintf("%s/.gcp/osServiceAccount.json", dirname))
+	command.Flags().StringVar(&credentialsFile, "credentials", defaultCredentials, "GCP credentials configuration file")
+}
+
+// RunOnGCP runs the given function on GCP, supplying it with a cloud instance connected to GCP and a reporter that writes to CLI.
+// The functions makes sure that infraID and region are specified, and extracts the credentials from a secret in order to connect to GCP.
+func RunOnGCP(gwInstanceType, kubeConfig, kubeContext string, dedicatedGWNodes bool,
+	function func(cloud api.Cloud, gwDeployer api.GatewayDeployer, reporter api.Reporter) error) error {
+	if ocpMetadataFile != "" {
+		err := initializeFlagsFromOCPMetadata(ocpMetadataFile)
+		utils.ExitOnError("Failed to read GCP Cluster information from OCP metadata file", err)
```

Comment:
The function signature returns an error and, here and below, in some cases we return an  error and others we `ExitOnError`. I think we should be consistent and do one or the other.

Reply:
I tried to follow the pattern used in AWS client code here - https://github.com/submariner-io/submariner-operator/blob/a68403944dd1643a894f37405b3db35269a425d3/pkg/subctl/cmd/cloud/aws/aws.go#L77

as well as here - https://github.com/submariner-io/submariner-operator/blob/a68403944dd1643a894f37405b3db35269a425d3/pkg/subctl/cmd/cloud/cleanup/aws.go#L44

If you are very particular, I can change it, no problem. Please let me know.

Reply:
I see. Regardless, I think we should be consistent either way. L96 is the only place where it propagates the error so we can change that to `ExitOnError`. Also that's the only place where it calls `reporter.Failed(err)` - I'm not sure if we need to notify the reporter but, again, we should be consistent. I assume we don't.

Reply:
Done